contracts/cw3-flex-multisig/src/state.rs

```diff
@@ -45,6 +86,57 @@ impl Proposal {
 
         status
     }
+
+    /// update_status sets the status of the proposal to current_status.
+    /// (designed for handler logic)
+    pub fn update_status(&mut self, block: &BlockInfo) {
+        self.status = self.current_status(block);
+    }
+
+    // returns true iff this proposal is sure to pass (even before expiration if no future
+    // sequence of possible votes can cause it to fail)
+    pub fn is_passed(&self, block: &BlockInfo) -> bool {
+        match self.threshold {
+            Threshold::AbsoluteCount {
+                weight: weight_needed,
+            } => self.votes.yes >= weight_needed,
+            Threshold::AbsolutePercentage {
+                percentage: percentage_needed,
+            } => self.votes.yes >= votes_needed(self.total_weight, percentage_needed),
+            Threshold::ThresholdQuora { threshold, quorum } => {
+                // this one is tricky, as we have two compares:
+                if self.expires.is_expired(block) {
+                    // * if we have closed yet, we need quorum% of total votes to have voted (counting abstain)
+                    //   and threshold% of yes votes from those who voted (ignoring abstain)
+                    let total = self.votes.total();
+                    let opinions = total - self.votes.abstain;
+                    total >= votes_needed(self.total_weight, quorum)
+                        && self.votes.yes >= votes_needed(opinions, threshold)
+                } else {
+                    // * if we have not closed yet, we need threshold% of yes votes (from 100% voters - abstain)
+                    //   as we are sure this cannot change with any possible sequence of future votes
+                    // * we also need quorum (which may not always be the case above)
+                    self.votes.total() >= votes_needed(self.total_weight, quorum)
+                        && self.votes.yes
+                            >= votes_needed(self.total_weight - self.votes.abstain, threshold)
+                }
```

Comment:
Both branches are the same. We can remove the expiration condition entirely.

Reply:
We can pull the quorum check out of this, you are correct there.

What is different is the total we compare against. Look closely at the total: unexpired is against the total number of votes (minus abstain), expired is against total weight (minus abstain)

Reply:
No, they are not, and this is an important difference.

Once the voting is over (expired), the passing threshold must be computed over the total weight of those **who voted**.

Reply:
Pushing a commit that makes this much more clear

Reply:
Crossed posts here. Good commit separating threshold from quorum checks.